cmd/notation/internal/plugin/plugin.go

```diff
@@ -0,0 +1,179 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/notaryproject/notation-go/dir"
+	"github.com/notaryproject/notation-go/log"
+	"github.com/notaryproject/notation-go/plugin"
+	"github.com/notaryproject/notation-go/plugin/proto"
+	notationerrors "github.com/notaryproject/notation/cmd/notation/internal/errors"
+	"github.com/notaryproject/notation/internal/osutil"
+	"github.com/notaryproject/notation/internal/trace"
+	"github.com/notaryproject/notation/internal/version"
+	"github.com/opencontainers/go-digest"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/mod/semver"
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+// maxPluginSourceBytes specifies the limit on how many response
+// bytes are allowed in the server's response to the download from URL request
+var maxPluginSourceBytes int64 = 256 * 1024 * 1024 // 256 MiB
+
+// PluginSourceType is an enum for plugin source
+type PluginSourceType int
+
+const (
+	// PluginSourceTypeFile means plugin source is file
+	PluginSourceTypeFile PluginSourceType = 1 + iota
+
+	// PluginSourceTypeURL means plugin source is URL
+	PluginSourceTypeURL
+)
+
+const (
+	// MediaTypeZip means plugin file is zip
+	MediaTypeZip = "application/zip"
+
+	// MediaTypeGzip means plugin file is gzip
+	MediaTypeGzip = "application/x-gzip"
+)
+
+// GetPluginMetadataIfExist returns plugin's metadata if it exists in Notation
+func GetPluginMetadataIfExist(ctx context.Context, pluginName string) (*proto.GetMetadataResponse, error) {
+	mgr := plugin.NewCLIManager(dir.PluginFS())
+	plugin, err := mgr.Get(ctx, pluginName)
+	if err != nil {
+		return nil, err
+	}
+	return plugin.GetMetadata(ctx, &proto.GetMetadataRequest{})
+}
+
+// GetPluginMetadata returns plugin's metadata given plugin path
+func GetPluginMetadata(ctx context.Context, pluginName, path string) (*proto.GetMetadataResponse, error) {
```

Comment:
This is because, `GetPluginMetadataIfExist` only checks if `plugin with name <pluginName>` is already installed in Notation.
However, `GetPluginMetadata` gets metadata of a plugin `candidate` at location `path`. At the same time, it expects the plugin name returned from metadata matches the input `<pluginName>`. If any step failed, `GetPluginMetadata` would return an error. This prevents installing a malformed plugin to Notation.

Reply:
Updated the code.